### PR TITLE
Use thinc optimizer in Shim classes

### PR DESCRIPTION
**Pull request by justindujardin**
Update the shim classes to use the thinc optimizers instead of creating framework-specific ones. 

- [x] Update tensorflow shim
- [x] Update mxnet shim
- [x] Update torch shim
- [x] Verify torch with cuda
- [x] Verify unique parameter ID use in torch/mxnet

### CODE REVIEW

File: thinc/shims/mxnet.py [hunk 1]
```diff
@@ -56,53 +56,14 @@ def backprop(grads):
...
-                ops.update_averages(sgd.averages[key], xp_param, sgd.nr_update[key])
-            else:
-                sgd.averages[key] = xp_param.copy()
-                sgd.nr_update[key] = init_steps
+    def finish_update(self, optimizer: Optimizer):
+        ctx = mx.current_context()
+        for value in self._model.collect_params().values():
+            grad = cast(FloatsXd, mxnet2xp(value.grad(ctx)))
+            param = cast(FloatsXd, mxnet2xp(value.data(ctx)))
+            param, _ = optimizer((value.name, value.name), param, grad)
```
Comment (justindujardin) [inline thinc/shims/mxnet.py:64]: I don't think the `(value.name, value.name)` tuples will be unique necessarily, and the "update_averages" code fails with shape errors if you have overlapping keys that don't match. I need to figure out how to get unique parameter IDs out of mxnet and pytorch

File: thinc/shims/pytorch.py [hunk 2]
```diff
@@ -53,39 +54,13 @@ def backprop(grads):
...
-                args["momentum"] = sgd.b1
-        if self._optimizer is None:
-            self._optimizer = cls(self._model.parameters(), **args)
-        else:
-            for param_group in self._optimizer.param_groups:
-                param_group.update(args)
-        return self._optimizer
+    def finish_update(self, optimizer: Optimizer):
+        for name, value in self._model.named_parameters():
+            cpu_data = cast(FloatsXd, torch2xp(value.data))
```
Comment (honnibal) [inline thinc/shims/pytorch.py:59]: Hm, `cpu_` seems like a misnomer, right? We're not copying between devices, we're just changing the container. We can call it `xp_data`, and maybe make the loop variable `torch_data`? Alternatively I've been using `array` for the numpy/cupy typed data and `tensor` for the pytorch-typed data, but that's maybe confusing.

**Comment by honnibal** [comment]
Looking good! Just that naming tweak and then we can go ahead and merge.

File: thinc/shims/pytorch.py (hunk 2 above)
Comment (justindujardin) [reply thinc/shims/pytorch.py]: Nice catch, you're right! I originally had this as `value.data.cpu().numpy()` before switching to torch2xp. xp_data sounds more appropriate now

**Pull request merged**